src/utilities/src/distance_function.C

```diff
@@ -0,0 +1,658 @@
+
+  // Add distance function variable
+  sys.add_variable("distance", libMesh::FIRST);
+
+  // Attach initialization function
+  sys.attach_init_object(*this);
+
+}
+
+//---------------------------------------------------
+// Compute distance function
+//
+void DistanceFunction::initialize ()
+{
+  // Call the compute function
+  this->compute();
+  return;
+}
+
+
+//---------------------------------------------------
+// Compute distance from input node to boundary_mesh
+//
+libMesh::Real DistanceFunction::node_to_boundary (const libMesh::Node* node)
+{
+  // Ensure that node is not NULL
+  libmesh_assert( node != NULL );
+
+  // Initialize distance to infinity
+  libMesh::Real distance = std::numeric_limits<libMesh::Real>::infinity();
+
+  // Get dimension
+  const unsigned int dim = _equation_systems.get_mesh().mesh_dimension();
+  libmesh_assert( (dim==2) || (dim==3) );
+
+  // Get coordinates of node
+  std::vector<libMesh::Real> xnode(dim);
+  for( unsigned int ii=0; ii<dim; ii++ ) xnode[ii] = (*node)(ii);
+
+  // This function will work on a distributed interior mesh, but won't
+  // give correct results on a distributed boundary mesh.
+  libmesh_assert(_boundary_mesh.is_serial());
+
+  // First loop over the boundary mesh and compute the distance to the
+  // nodes.  This will give us an idea of what elements to check more
+  // closely.
+
+  libMesh::MeshBase::const_element_iterator       el     = _boundary_mesh.active_elements_begin();
+  const libMesh::MeshBase::const_element_iterator end_el = _boundary_mesh.active_elements_end();
+
+  if (el==end_el)
+    {
+      std::cout << "There are no elements to iterate through!!!  Returning..." << std::endl;
+      return distance;
+    }
+
+  const libMesh::Elem* elem_containing_min = NULL;
+  libMesh::Real dmin = std::numeric_limits<Real>::infinity();
+
+  for ( ; el != end_el; ++el) {
+
+    const libMesh::Elem* belem = *el;
+
+    for (unsigned int bnode=0; bnode<belem->n_nodes(); ++bnode)
+      {
+	const libMesh::Point& pbndry = belem->point(bnode);
+
+	libMesh::Real dnode = 0.0;
+	for (unsigned int idim=0; idim<dim; ++idim)
+	  dnode += (xnode[idim] - pbndry(idim))*(xnode[idim] - pbndry(idim));
+
+	dnode = std::sqrt(dnode);
+
+	if (dnode<dmin)
+	  {
+	    dmin = dnode;
+	    elem_containing_min = belem;
+	  }
+      }
+
+  } // end element loop
+
+  // grab the elements around the minimum
+  std::set<const libMesh::Elem*> near_min_elems;
+  elem_containing_min->find_point_neighbors (near_min_elems);
+
+  // insert the element itself into the set... we must check it also
+  near_min_elems.insert(elem_containing_min);
+
+  std::set<const libMesh::Elem*>::iterator nm_el;
+
+  // loop over the near_min_elems
+  for (nm_el=near_min_elems.begin(); nm_el!=near_min_elems.end(); ++nm_el)
+    {
+
+      const libMesh::Elem* belem = *nm_el; //*el;
+
+      // Ensure that elem defined by edge/face is linear
```

Comment:
You ought to just add the missing EDGE3/TRI6/QUAD9 cases here, but with (at least for the first two) a has_affine_map() assertion.  That'll at least cover the "geometrically first order, but higher order FE" cases with very little work.